### PR TITLE
[SIMT] Add ballot_sync warp intrinsics

### PR DESCRIPTION
**Pull request by Wimacs**
Related issue = #4631 
Add `ballot_sync` intrinsics for cuda backend.
According to #4631, I added few python codes to call intrinsics, but have some problems when running the tests. In NVVM IR `nvvm_vote_ballot` needs to accept a bool `i1` parameter, but it seems that the frontend can only generate a `(i32, i32)->i32` func, it should be `(i32, i1)->i32`. Is there any hint or solution to generate `i1` variable or transfer `i32` to `i1`? Or I miss some points hope some guys can help @yuanming-hu .

### CODE REVIEW

**Comment by yuanming-hu** [comment]
Ah you find something interesting in Taichi. IIRC `u1` has not been supported - we use `u32/i32` for boolean. (@strongoier a good thing to cover in your new type system :-) Related issue: https://github.com/taichi-dev/taichi/issues/577

For now, maybe we can implement 
```C++
int32 cuda_ballot_i32(int32 input) {
  return cuda_ballot((bool)input);
}
```
 around 

https://github.com/taichi-dev/taichi/blob/c902db6236648d938656c2ededc8062284da3291/taichi/runtime/llvm/runtime.cpp#L1011-L1013

and call that function instead.

**Comment by qiao-bo** [comment]
Hi, thanks for contributing. Can you please fix the conflict ;)?

File: taichi/runtime/llvm/runtime.cpp [hunk 1]
```diff
@@ -1040,6 +1040,14 @@ int32 cuda_ballot_sync(int32 mask, bool bit) {
   return 0;
 }
 
+int32 cuda_ballot_i32(int32 predicate) {
+  return cuda_ballot_sync(UINT32_MAX, (bool)predicate);
+}
+
+int32 cuda_ballot_sync_i32(u32 mask, int32 predicate) {
```
Comment (qiao-bo) [inline taichi/runtime/llvm/runtime.cpp:1047]: do we need and use this function?

Comment (Wimacs) [reply taichi/runtime/llvm/runtime.cpp:1047]: not yet,  in #4631 it has not yet been decided whether to block the mask or not, so both of them are temporarily implemented

File: tests/python/test_simt.py [hunk 2]
```diff
@@ -24,8 +26,24 @@ def test_unique():
...
+
+    @ti.kernel
+    def foo():
+        ti.loop_config(block_dim=32)
+        for i in range(32):
+            a[i] = ti.simt.warp.ballot(b[i])
+
+    key = 0
+    for i in range(32):
+        b[i] = random.randint(1, 100) % 2
```
Comment (qiao-bo) [inline tests/python/test_simt.py:40]: nit, can you change this to utilize ti.random()? ;/

Comment (yuanming-hu) [reply tests/python/test_simt.py:40]: My two cents: use constant numbers instead of random numbers to avoid random test behaviors.

**Review by yuanming-hu — APPROVED** [review]
Awesome, LGTM!

**Review by qiao-bo — APPROVED** [review]
Thanks!

**Pull request merged**